src/metabase/api/docs.clj

```diff
@@ -1,13 +1,97 @@
+
+(def openapi-file-path
+  "Path to the local OpenAPI specification file."
+  "resources/openapi/openapi.json")
+
+(defn- sort-keys
+  "Sort maps and sets alphabetically to reduce diff noise on openapi.json"
+  [data]
+  (walk/postwalk
+   (fn [x]
+     (cond
+       (map? x) (into (sorted-map) x)
+       (set? x) (into (sorted-set) x)
+       :else    x))
+   data))
+
+(defn write-openapi-spec-to-file!
+  "Generate and write the OpenAPI specification to a local file.
+  Takes the root handler and generates the complete OpenAPI spec, writing it to [[openapi-file-path]]."
+  [root-handler]
+  (try
+    (let [spec (merge
+                (open-api/root-open-api-object root-handler)
+                {:servers [{:url         ""
+                            :description "Metabase API"}]})
+          file (io/file openapi-file-path)]
+      ;; Create parent directory if it doesn't exist
+      (when-let [parent-dir (.getParentFile file)]
+        (.mkdirs parent-dir))
+      (json/encode-to (sort-keys spec) (io/writer file) {:pretty true})
+      (log/info "OpenAPI specification written to" openapi-file-path))
+    (catch Throwable e
+      (log/error e "Failed to write OpenAPI specification to file"))))
+
+(defonce ^:private openapi-regen-state
+  (atom {:executing? false
+         :needs-regen? false}))
+
+(def ^:private debounce-delay-ms
+  "Delay in milliseconds before starting regeneration to allow multiple rapid requests to coalesce."
+  200)
+
+(defn request-spec-regeneration!
+  "Request OpenAPI spec regeneration. Multiple rapid requests are coalesced into one execution.
+  Uses a debounce delay to batch requests that arrive in quick succession (e.g., when re-evaling a file with multiple endpoints)."
+  [routes]
+  (swap! openapi-regen-state assoc :needs-regen? true)
+
+  (when-not (:executing? @openapi-regen-state)
+    (swap! openapi-regen-state assoc :executing? true)
+    (future
+      (try
+        ;; Wait for debounce period to let multiple requests coalesce
+        (Thread/sleep ^Long debounce-delay-ms)
+        (loop []
+          (when (:needs-regen? @openapi-regen-state)
+            (swap! openapi-regen-state assoc :needs-regen? false)
+            (log/info "Regenerating OpenAPI specification...")
+            (write-openapi-spec-to-file! routes)
+            (log/info "OpenAPI specification regenerated successfully")
+            ;; Check if more requests came in while we were working
+            (recur)))
+        (catch Throwable e
+          (log/error e "Error regenerating OpenAPI specification"))
+        (finally
+          (swap! openapi-regen-state assoc :executing? false))))))
+
+(defn- read-openapi-spec-from-file
+  "Read the OpenAPI specification from the local file.
+  Returns the parsed JSON content, or nil if the file doesn't exist or can't be read."
+  []
+  (try
+    (let [file (io/file openapi-file-path)]
+      (when (.exists ^java.io.File file)
+        (json/decode+kw (io/reader file))))
```

Comment:
I think we want to use with-open here since `json/decode+kw` will read from the stream but won’t close it. Leaving readers open can hold file handles and cause locking issues on some platforms